crates/wasm_interp/src/execute.rs

```diff
@@ -197,26 +257,79 @@ impl<'a> ExecutionState<'a> {
             }
             NOP => {}
             BLOCK => {
-                self.block_depth += 1;
-                todo!("{:?} @ {:#x}", op_code, file_offset);
+                self.fetch_immediate_u32(module); // blocktype (ignored)
+                self.block_loop_addrs.push(None);
             }
             LOOP => {
-                self.block_depth += 1;
-                todo!("{:?} @ {:#x}", op_code, file_offset);
+                self.fetch_immediate_u32(module); // blocktype (ignored)
+                self.block_loop_addrs
+                    .push(Some(self.program_counter as u32));
+            }
+            IF => {
+                self.fetch_immediate_u32(module); // blocktype (ignored)
+                let condition = self.value_stack.pop_i32();
+                self.block_loop_addrs.push(None);
+                if condition == 0 {
+                    let mut depth = self.block_loop_addrs.len();
+                    loop {
+                        let skipped_op = OpCode::from(module.code.bytes[self.program_counter]);
+                        OpCode::skip_bytes(&module.code.bytes, &mut self.program_counter).unwrap();
+                        match skipped_op {
+                            BLOCK | LOOP | IF => {
+                                depth += 1;
+                            }
+                            END => {
+                                depth -= 1;
+                            }
+                            ELSE => {
+                                if depth == self.block_loop_addrs.len() {
+                                    break;
+                                }
+                            }
+                            _ => {}
+                        }
+                    }
+                }
+            }
+            ELSE => {
+                // We only reach this point when we finish executing the "then" block of an IF statement
+                // (For a false condition, we would have skipped past the ELSE when we saw the IF)
+                // We don't want to execute the ELSE block, so we skip it, just like `br 0` would.
+                self.do_break(0, module);
             }
-            IF => todo!("{:?} @ {:#x}", op_code, file_offset),
-            ELSE => todo!("{:?} @ {:#x}", op_code, file_offset),
             END => {
-                if self.block_depth == 0 {
+                if self.block_loop_addrs.is_empty() {
```

Comment:
Oops just realised this condition is wrong! `block_loop_addrs` contains data for the entire call stack. It is only empty when the whole call stack is finished, not just the current function. We just don't have a test that would catch it. Yikes!

Reply:
Pushed a fix for this.